MovieTime/MovieTime/Service/MTParameterEncoding.swift

```diff
@@ -0,0 +1,375 @@
+        var url = url
+        if customize == .reflection {
+            // Encodes parameters in URL
+            if method.encodesParametersInURL() {
+                url = url.reflectionEncode(parameters: parameters)
+            }
+        }
+        guard let encodedURL = url.addingPercentEncoding(withAllowedCharacters: NSCharacterSet.urlQueryAllowed) else { return nil }
+
+        return URL(string: encodedURL)
+    }
+
+    public func encode(_ urlRequest: URLRequest, with parameters: MTParameters?) -> URLRequest {
+        var urlRequest = urlRequest
+
+        guard let parameters = parameters, let array = parameters[MTArrayParametersKey] else {
+            return urlRequest
+        }
+        guard let method = MTHTTPMethod(rawValue: urlRequest.httpMethod ?? "GET"), !method.encodesParametersInURL() else {
+            return urlRequest
+        }
+
+        do {
+            let data = try JSONSerialization.data(withJSONObject: array, options: options)
+
+            if urlRequest.value(forHTTPHeaderField: "Content-Type") == nil {
+                urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
+            }
+
+            urlRequest.httpBody = data
+        } catch {
+            return urlRequest
+        }
+
+        return urlRequest
+    }
+}
+
+// MARK: - Custom Extensions
+
+/// Extension that allows an array be sent as a request parameters
+extension Array {
+    func asParameters() -> MTParameters {
+        return [MTArrayParametersKey: self]
+    }
+}
+
+extension NSArray {
+    func asParameters() -> MTParameters {
+        return [MTArrayParametersKey: self]
+    }
+}
+
+extension String {
+
+    func reflectionEncode(parameters: MTParameters?) -> String {
+        guard let parameters = parameters else { return self }
+
+        var strUrl = self
+
+        for (key, value) in parameters {
+
+            let strKey = String(describing: key)
+            var strValue = ""
+
+            if let array = value as? [Any] {
+
+                // For array objects
+                for (index, arrayValue) in array.enumerated() {
+
+                    let currentValue = String(describing: arrayValue)
+                    strValue = strValue.appendingFormat("%@=%@", strKey, currentValue)
+                    if index != (array.count - 1) {
+                        strValue = strValue.appending("&")
+                    }
+                }
+            } else {
+
+                // For any objects
+                strValue = String(describing: value)
+            }
+
+            strUrl = strUrl.replacingOccurrences(of: "{" + strKey + "}", with: strValue)
+        }
+
+        return strUrl
+    }
+}
+
+extension NSNumber {
+    fileprivate var isBool: Bool { return CFBooleanGetTypeID() == CFGetTypeID(self) }
+}
+
+extension MTHTTPMethod {
+    fileprivate func encodesParametersInURL() -> Bool {
```

Comment:
could you explain how do you to decide between creating a method or a computed property?

Reply:
During development, I forgot that enum extensions can contain computed properties. Even the enum can't contain stored properties a computed property is better in this context.